code/object/objcollide.cpp

```diff
@@ -607,23 +608,25 @@ void obj_reset_colliders()
 
 void obj_collide_retime_cached_pairs()
 {
-	for ( auto& pair : Collision_cached_pairs ) {
-		pair.second.next_check_time = timestamp(0);
+	TRACE_SCOPE(tracing::RetimeCollisionCache);
+
+	auto it = Collision_cached_pairs.begin();
+	while (it != Collision_cached_pairs.end()) {
+		auto &pair = it->second;
+		if (pair.signature_a != pair.a->signature || pair.signature_b != pair.b->signature) {
+			it = Collision_cached_pairs.erase(it);
+		} else {
+			if (pair.a->flags[Object::Object_Flags::Collision_cache_stale] || pair.b->flags[Object::Object_Flags::Collision_cache_stale])
```

Comment:
The flag should be cleared on the stale object.

Reply:
Whoops!

Reply:
Eugh, crap now we're back to the issues with object flags, it's difficult to know while we're iterating when we're done with an object and to clear its flag.

Reply:
It's not the prettiest but may as well bring back the list of stale cache objects.

Reply:
Oh, right, because the stale object may have more than one collision pair.  That definitely makes it more complicated.